remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteStorageManager.java

```diff
@@ -0,0 +1,171 @@
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class is an implementation of {@link RemoteStorageManager} backed by inmemory store.
+ */
+public class InmemoryRemoteStorageManager implements RemoteStorageManager {
+    private static final Logger log = LoggerFactory.getLogger(InmemoryRemoteStorageManager.class);
+
+    // map of key to log data, which can be segment or any of its indexes.
+    private Map<String, byte[]> keyToLogData = new ConcurrentHashMap<>();
+
+    public InmemoryRemoteStorageManager() {
+    }
+
+    static String generateKeyForSegment(RemoteLogSegmentMetadata remoteLogSegmentMetadata) {
+        return remoteLogSegmentMetadata.remoteLogSegmentId().id().toString() + ".segment";
+    }
+
+    static String generateKeyForIndex(RemoteLogSegmentMetadata remoteLogSegmentMetadata,
+                                      IndexType indexType) {
+        return remoteLogSegmentMetadata.remoteLogSegmentId().id().toString() + "." + indexType.toString();
+    }
+
+    // visible for testing.
+    boolean containsKey(String key) {
+        return keyToLogData.containsKey(key);
+    }
+
+    @Override
+    public void copyLogSegmentData(RemoteLogSegmentMetadata remoteLogSegmentMetadata,
+                                   LogSegmentData logSegmentData)
+            throws RemoteStorageException {
+        log.debug("copying log segment and indexes for : {}", remoteLogSegmentMetadata);
+        Objects.requireNonNull(remoteLogSegmentMetadata, "remoteLogSegmentMetadata can not be null");
+        Objects.requireNonNull(logSegmentData, "logSegmentData can not be null");
+        try {
+            keyToLogData.put(generateKeyForSegment(remoteLogSegmentMetadata),
+                    Files.readAllBytes(logSegmentData.logSegment().toPath()));
+            keyToLogData.put(generateKeyForIndex(remoteLogSegmentMetadata, IndexType.Offset),
+                    Files.readAllBytes(logSegmentData.offsetIndex().toPath()));
+            keyToLogData.put(generateKeyForIndex(remoteLogSegmentMetadata, IndexType.Timestamp),
+                    Files.readAllBytes(logSegmentData.timeIndex().toPath()));
+            keyToLogData.put(generateKeyForIndex(remoteLogSegmentMetadata, IndexType.Transaction),
+                    Files.readAllBytes(logSegmentData.txnIndex().toPath()));
+            keyToLogData.put(generateKeyForIndex(remoteLogSegmentMetadata, IndexType.LeaderEpoch),
+                    logSegmentData.leaderEpochIndex().array());
+            keyToLogData.put(generateKeyForIndex(remoteLogSegmentMetadata, IndexType.ProducerSnapshot),
+                    Files.readAllBytes(logSegmentData.producerSnapshotIndex().toPath()));
+        } catch (IOException e) {
+            throw new RemoteStorageException(e.getMessage(), e);
```

Comment:
We could add a c'tor overload to [RemoteStorageException](https://github.com/apache/kafka/blob/0d9a95a7d0ab06aecc4480901707e29dd2a3147e/clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageException.java) that takes a `Throwable` as argument, it would the need to pass 2 args here.